src/corelibs/U2View/src/ov_sequence/view_rendering/SequenceViewAnnotatedRenderer.cpp

```diff
@@ -188,16 +187,56 @@ void SequenceViewAnnotatedRenderer::drawAnnotation(QPainter& p, const QSize& can
                     const QString aText = prepareAnnotationText(aData, as);
                     drawBoundedText(p, annotationRect, aText);
                 }
-                if (simple &&
-                    annotationRect.width() > MIN_WIDTH_TO_DRAW_EXTRA_FEATURES &&
-                    displaySettings.displayCutSites) {
-                    drawCutSite(p, aData, r, annotationRect, as->color, canvasSize, visibleRange);
-                }
             }
             drawAnnotationConnections(p, a, as, displaySettings, canvasSize, visibleRange);
-        } else if (isRestrictionSite && simple && displaySettings.displayCutSites) {
-            p.setPen(borderPen);
-            drawCutSite(p, aData, r, QRect(), as->color, canvasSize, visibleRange);
+        }
+    }
+    // draw restriction sites
+    /**
+     * Calculates restriction cuts positions depends on visible area and calls drawCutSite.
+     *
+     * \param reg Region of the annotation, which sites should be drawn.
+     * \param drawDirect Draw direct cut if true and do not draw if false.
+     * \param drawComplement Draw complement cut if true do not draw if false.
+     */
+    auto drawCutSiteLambda = [this, &p, &visibleRange, &a, &as, &availableHeight, &canvasSize, selected, &aData](const U2Region& reg, bool drawDirect = true, bool drawComplement = true) {
+        if (reg.intersects(visibleRange)) {
+
+            const U2Region visibleLocation = reg.intersect(visibleRange);
+            const U2Region y = getAnnotationYRange(a, 0, as, availableHeight);
+            CHECK(y.startPos >= 0, );
+
+            const int x1 = posToXCoord(visibleLocation.startPos, canvasSize, visibleRange);
+            const int x2 = posToXCoord(visibleLocation.endPos(), canvasSize, visibleRange);
+
+            const int rw = qMax(selected ? MIN_SELECTED_ANNOTATION_WIDTH : MIN_ANNOTATION_WIDTH, x2 - x1);
+            SAFE_POINT(rw > 0, "Negative length of annotationYRange", );
+
+            const QRect annotationRect(x1, y.startPos, rw, y.length);
+            if (annotationRect.width() > MIN_WIDTH_TO_DRAW_EXTRA_FEATURES) {
+                drawCutSite(p, aData, reg, annotationRect, as->color, canvasSize, visibleRange, drawDirect, drawComplement);
+            }
+        } else {
+            drawCutSite(p, aData, reg, QRect(), as->color, canvasSize, visibleRange);
+        }
+    };
+
+    if (isRestrictionSite && displaySettings.displayCutSites) {
+        p.setPen(borderPen);
+        const bool isCircular = ctx->getSequenceObject()->isCircular();
+        auto seqLength = ctx->getSequenceLength();
+        auto merged = U1AnnotationUtils::mergeAnnotatedRegionsAroundJunctionPoint(location, seqLength);
```

Comment:
Move circular sequence related login under `isCircular` check. No need  to calculate merged regions for non-circular sequences

Reply:
OK